Adding query metrics to the Oracle DBM checks

nenadnoveljic opened the pull request:
### What does this PR do?

We're adding query metrics to the Oracle DBM check.

### Motivation

Query metrics is a standard DBM feature that is also implemented for other database products.

### Additional Notes

The Oracle check collects much more information for a captured SQL statement than it's the case with SQL Server, Postgres and MySQL. The metrics are read from the [V$SQLSTATS](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/V-SQLSTATS.html#GUID-495DD17D-6741-433F-871D-C965EB221DA9) view.

### Possible Drawbacks / Trade-offs

Query metrics are collected for the queries captured in two consecutive activity snapshots. We didn't want to scan the query cache as this strategy was causing performance bottlenecks in the SQL Server and Postgres integrations. In other words, the primary concern was not to cause any problems on customer's databases. On the downside, we're collecting query metrics for less queries.  If a customer would like to get more granular data and is willing to accept a higher load, we can implement a second strategy with a separate change. 

### Describe how to test/QA your changes

1. Generate load
2. Verify that the query metrics are shown in the DBM GUI.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

nenadnoveljic committed:
test cases

nenadnoveljic committed:
commented out win tests

nenadnoveljic committed:
dbm-oracle-beta-0.21

nenadnoveljic committed:
fqt

nenadnoveljic committed:
[WIP] query metrics

nenadnoveljic committed:
[WIP] query metrics

nenadnoveljic committed:
completed

nenadnoveljic committed:
include DD queries

nenadnoveljic committed:
port & query metrics default true

nenadnoveljic committed:
renamed conf file

nenadnoveljic committed:
release notes added

nenadnoveljic committed:
deleted conf.yaml.default

nenadnoveljic committed:
Merge branch 'main' into nenad.noveljic/oracle-query-metrics

jmeunier28 commented on cmd/agent/dist/conf.d/oracle-dbm.d/conf.yaml.example:
```diff
@@ -15,103 +15,121 @@ instances:
+      ## @param enabled - boolean - optional - default: true
+      ## Enable collection of query metrics. Requires `dbm: true`.
+      #
+      # enabled: true
```
this looks incorrect when comparing to how you have [defined this struct](https://github.com/DataDog/datadog-agent/pull/17000/files#diff-365baa745bde9428c3cd5738c844f63061a48ccfa1848b48bfc6f4d20176bfdaR40)

should this be:
```
QueryMetrics map[string]QueryMetricConfig `yaml:"query_metrics"`
...
...

type QueryMetricConfig struct {
Enabled bool `yaml:"enabled"`
}
```

nenadnoveljic replied:
Corrected now, thx. I also added the `query_samples` parameter to control sampling.

nenadnoveljic replied:
> so if we want to be consistent across our integrations the config query_samples refers to execution plan collection and query_activity refers to activity/load sampling
> 
> Maybe it makes sense to do something different here with the new integration, like:
> 
> query_samples -> activity/load samples
> collect_execution_plans -> plan collections
> idk, what do you think?

Using `query_samples` for controlling execution plans goes agains my intuition. "Sample" is the front end annotation for query samples, so `query_samples` is better aligned with the GUI:

![image](https://github.com/DataDog/datadog-agent/assets/18366081/dc86ce72-9285-4d1f-8037-9e79690643fb)

Also @justiniso recognised the inconsistent naming and [wants to change it](https://dd.slack.com/archives/C027TGQFUMR/p1683224818003169?thread_ts=1682425251.404039&cid=C027TGQFUMR)

I have already introduced the parameter `execution_plans` for controlling the collection of execution plans. I could change it to `collect_execution_plans` if that's more appropriate. That parameter will be introduced in a [separate branch](https://github.com/DataDog/datadog-agent/blob/nenad.noveljic/oracle-execution-plans/pkg/collector/corechecks/oracle-dbm/config/config.go#L42) that implements only execution plan collection functionality.

nenadnoveljic committed:
config alignment to other db integrations

nenadnoveljic committed:
config example correction

jmeunier28 commented:
> Corrected now, thx. I also added the query_samples parameter to control sampling.

cool! so if we want to be consistent across our integrations the config `query_samples` refers to execution plan collection and `query_activity` refers to activity/load sampling

Maybe it makes sense to do something different here with the new integration, like:

```
query_samples -> activity/load samples
collect_execution_plans -> plan collections
```

idk, what do you think?

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/oracle_test.go:
```diff
@@ -88,7 +88,8 @@ func TestConnection(t *testing.T) {
 
 func demuxOpts() aggregator.AgentDemultiplexerOptions {
-	opts := aggregator.DefaultAgentDemultiplexerOptions(nil)
+	//opts := aggregator.DefaultAgentDemultiplexerOptions(nil)
```
meant to remove this?

nenadnoveljic replied:
done

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/oracle_test.go:
```diff
@@ -119,8 +120,14 @@ func connectToDB(driver string) (*sqlx.DB, error) {
 func TestChkRun(t *testing.T) {
-	aggregator.InitAndStartAgentDemultiplexer(demuxOpts(), "")
+	//aggregator.InitAndStartAgentDemultiplexer(demuxOpts(), "")
+	//aggregator.InitAndStartAgentDemultiplexer(nil, demuxOpts(), "")
```
can we remove these comments now?

nenadnoveljic replied:
done

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -202,39 +230,40 @@ func ConstructStatementMetricsQueryBlock(sqlHandleColumn string, whereClause str
-		}
+			} else {
+				return nil, fmt.Errorf("statements wrong driver %s", driver)
+			} */
```
can we clean this up or is this comment needed?

nenadnoveljic replied:
done

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -267,8 +296,32 @@ func (c *Check) StatementMetrics() (int, error) {
+			var DDForceMatchingSignatures []string
+			err = c.db.Select(
+				&DDForceMatchingSignatures,
+				"SELECT distinct force_matching_signature FROM v$sqlstats WHERE sql_text like '%/* DD%' and (sysdate-last_active_time)*3600*24 < :1",
```
is there a better way we can get these queries? like can we just filter for whatever user they configure for the ddagent? Or are you concerned that won't be distinct enough?

nenadnoveljic replied:
> like can we just filter for whatever user they configure for the ddagent? Or are you concerned that won't be distinct enough?
[V$SQLSTATS](https://docs.oracle.com/en/database/oracle/oracle-database/19/refrn/V-SQLSTATS.html#GUID-495DD17D-6741-433F-871D-C965EB221DA9) doesn't contain the information about user, so we can filter on user. It doesn't contain any information about session related data either, so by we can't tell who was executing this query.

> is there a better way we can get these queries?
Filtering on the comment text seems straightforward to me. What are your concerns?

jmeunier28 replied:
It just feels a bit hacky, but if we can't look up the user then that seems fine

nenadnoveljic replied:
The technical reason is that the SQL statements and their execution statistics are kept in cache, so the same statement can be run by many users and sessions. The accounting is done on the SQL cursor level.

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -194,6 +195,33 @@ type MetricsPayload struct {
+	Source       string      `json:"ddsource"`
+	Tags         []string    `json:"tags,omitempty"`
+	DBMType      string      `json:"dbm_type"`
+	FQTDB        FQTDB       `json:"db"`
```
for the future: FQT events should follow the same exact structure [as explain events](https://github.com/DataDog/dd-go/blob/prod/database-monitoring/apps/dbm-events-processor/decoder/payloads.go#L302-L317) (minus just a few fields, of course). It probably makes sense to define a shared generic struct that we can use between this and the thread that is collecting execution plans

nenadnoveljic replied:
Created a [Jira ticket](https://datadoghq.atlassian.net/browse/DBM-1595) for that.

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -398,51 +452,96 @@ func (c *Check) StatementMetrics() (int, error) {
+							commands:       obfuscatedStatement.Commands,
+						}
+						if statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature != "" {
+							c.DDstatementsCache.forceMatchingSignatures[statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature] = cacheEntry
```
is this the correct cache to store this new entry in? I would think it should be this one `c.statementsCache.forceMatchingSignatures`?

jmeunier28 commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -194,6 +195,33 @@ type MetricsPayload struct {
+	Host         string      `json:"host,omitempty"` // Host is the database hostname, not the agent hostname
+	AgentVersion string      `json:"ddagentversion,omitempty"`
+	Source       string      `json:"ddsource"`
+	Tags         []string    `json:"tags,omitempty"`
```
This should be `ddtags` and a `string` instead of a slice of strings. I believe right now this field is just being ignored 

here is what we are expecting [in the backend](https://github.com/DataDog/dd-go/blob/prod/database-monitoring/apps/dbm-events-processor/decoder/payloads.go#L304)

nenadnoveljic replied:
converted it to comma separated tags & changed the json field name to `ddtags`

nenadnoveljic commented:
> > Corrected now, thx. I also added the query_samples parameter to control sampling.
> 
> cool! so if we want to be consistent across our integrations the config `query_samples` refers to execution plan collection and `query_activity` refers to activity/load sampling
> 
> Maybe it makes sense to do something different here with the new integration, like:
> 
> ```
> query_samples -> activity/load samples
> collect_execution_plans -> plan collections
> ```
> 
> idk, what do you think?

Using `query_samples` for controlling execution plans goes agains my intuition. "Sample" is the front end annotation for query samples, so `query_samples` is better aligned with the GUI:

![image](https://github.com/DataDog/datadog-agent/assets/18366081/dc86ce72-9285-4d1f-8037-9e79690643fb)

Also @justiniso recognised the inconsistent naming and [wants to change it](https://dd.slack.com/archives/C027TGQFUMR/p1683224818003169?thread_ts=1682425251.404039&cid=C027TGQFUMR)

I have already introduced the parameter `execution_plans` for controlling the collection of execution plans. I could change it to `collect_execution_plans` if that's more appropriate. That parameter will be introduced in a [separate branch](https://github.com/DataDog/datadog-agent/blob/nenad.noveljic/oracle-execution-plans/pkg/collector/corechecks/oracle-dbm/config/config.go#L42) that implements only execution plan collection functionality.

nenadnoveljic commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -267,8 +296,32 @@ func (c *Check) StatementMetrics() (int, error) {
+	totalSQLTextTimeUs := int64(0)
 	var oracleRows []OracleRow
-	if c.config.QueryMetrics {
+	FQTSent := make(map[string]int)
```
moved it downwards - just before the loop.

nenadnoveljic commented on pkg/collector/corechecks/oracle-dbm/statements.go:
```diff
@@ -398,51 +452,96 @@ func (c *Check) StatementMetrics() (int, error) {
+					SQLStatement = obfuscatedStatement.statement
+					found = true
+				} else if c.config.InstanceConfig.QueryMetrics.IncludeDatadogQueries {
+					obfuscatedStatement, ok := c.DDPrevStatementsCache.forceMatchingSignatures[statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature]
```
The "regular" cache, i.e. `c.statementsCache`, is filled by query samples. The agent is collecting query metrics for the queries captured in two consecutive samples. Since the agent statements are excluded from sampling, they would never land in this cache. For this reason, I introduced a separate cache for the agent statements.

jmeunier28 replied:
> Since the agent statements are excluded from sampling 

are all of the agent statements excluded from statements or just the load sampling query? It feels like we _should_ probably collect all of the agent statement that are _not_ the current session aka the sampling query as part of the activity sampling, no?

nenadnoveljic committed:
review

nenadnoveljic committed:
tags as string

jmeunier28 commented:
> Using query_samples for controlling execution plans goes agains my intuition. "Sample" is the front end annotation for query samples, so query_samples is better aligned with the GUI

yep totally agree! `execution_plans` sounds good to me!
